awx/ui/client/src/users/users-tokens-list.controller.js

```diff
@@ -0,0 +1,119 @@
+/** ***********************************************
+ * Copyright (c) 2018 Ansible, Inc.
+ *
+ * All Rights Reserved
+ ************************************************ */
+function ListTokensController (
+    $filter,
+    $scope,
+    $state,
+    Dataset,
+    strings,
+    ProcessErrors,
+    Rest,
+    GetBasePath,
+    Prompt,
+    Wait
+) {
+    const vm = this || {};
+
+    vm.strings = strings;
+    vm.activeId = $state.params.token_id;
+
+    $scope.canAdd = true;
+
+    // smart-search
+    const name = 'tokens';
+    const iterator = 'token';
+    const key = 'token_dataset';
+
+    $scope.list = { iterator, name, basePath: 'tokens' };
+    $scope.collection = { iterator };
+    $scope[key] = Dataset.data;
+    vm.tokensCount = Dataset.data.count;
+    $scope[name] = Dataset.data.results;
+    $scope.$on('updateDataset', (e, dataset) => {
+        $scope[key] = dataset;
+        $scope[name] = dataset.results;
+        vm.tokensCount = dataset.count;
+    });
+
+    vm.getLastUsed = token => {
+        const lastUsed = _.get(token, 'last_used');
+
+        if (!lastUsed) {
+            return undefined;
+        }
+
+        let html = $filter('longDate')(lastUsed);
+
+        const { username, id } = _.get(user, 'summary_fields.last_used', {});
+
+        if (username && id) {
+            html += ` by <a href="/#/users/${id}">${$filter('sanitize')(username)}</a>`;
+        }
+
+        return html;
+    };
+
+    vm.deleteToken = (tok) => {
+        const action = () => {
+            $('#prompt-modal').modal('hide');
+            Wait('start');
+            Rest.setUrl(`${GetBasePath('tokens')}${tok.id}`);
+            Rest.destroy()
+                .then(() => {
+                    let reloadListStateParams = null;
+
+                    if ($scope.tokens.length === 1 && $state.params.token_search &&
+                    !_.isEmpty($state.params.token_search.page) &&
+                    $state.params.token_search.page !== '1') {
+                        const page = `${(parseInt(reloadListStateParams
+                            .token_search.page, 10) - 1)}`;
+                        reloadListStateParams = _.cloneDeep($state.params);
```

Comment:
This would be more easily unit tested as a separate function that is called by `deleteToken`